rules/namespace.go

```diff
@@ -165,3 +308,71 @@ func (nss Namespaces) Schema() (*schema.Namespaces, error) {
 
 	return res, nil
 }
+
+// Namespace returns a namespace with a given name.
+func (nss *Namespaces) Namespace(name string) (*Namespace, error) {
+	var res *Namespace
+
+	for i, ns := range nss.namespaces {
+		if string(ns.name) != name {
+			continue
+		}
+
+		if res == nil {
+			res = &nss.namespaces[i]
+		} else {
+			return nil, errMultipleNamespaceMatches
+		}
+	}
+
+	if res == nil {
+		return nil, errNamespaceNotFound
+	}
+
+	return res, nil
+}
+
+// AddNamespace adds a new namespace to the namespaces structure and persists it
+func (nss *Namespaces) AddNamespace(nsName string) error {
+	existing, err := nss.Namespace(nsName)
+	if err != nil && err != errNamespaceNotFound {
+		return fmt.Errorf(namespaceActionErrorFmt, "add", nsName, err)
+	}
+
+	// Brand new namespace
+	if err == errNamespaceNotFound {
+		ns := Namespace{
+			name: []byte(nsName),
+			snapshots: []NamespaceSnapshot{
+				NamespaceSnapshot{
+					forRuleSetVersion: 1,
+					tombstoned:        false,
+				},
+			},
+		}
+
+		nss.namespaces = append(nss.namespaces, ns)
+		return nil
+	}
+
+	// Revive the namespace
+	if err = existing.revive(); err != nil {
+		return fmt.Errorf(namespaceActionErrorFmt, "revive", nsName, err)
+	}
+
+	return nil
+}
+
+// DeleteNamespace tombstones the given namespace mapping it to the given RuleSet version + 1
+func (nss *Namespaces) DeleteNamespace(nsName string, forRuleSetVersion int) error {
+	existing, err := nss.Namespace(nsName)
+	if err != nil {
+		return fmt.Errorf(namespaceActionErrorFmt, "delete", nsName, err)
+	}
+
+	if err := existing.markTombstoned(forRuleSetVersion + 1); err != nil {
```

Comment:
Just to confirm, this means the ruleset version saved in the namespace is the same as the version of the ruleset after the ruleset is persisted yeah?